### PR TITLE
Use official Docker GitHub actions

### PR DESCRIPTION
**Pull request by fauust**
Use the following actions instead of manual Docker buildx setup:
- https://github.com/docker/login-action
- https://github.com/docker/setup-qemu-action
- https://github.com/docker/build-push-action

Notes:
- It would be good to merge "Build and push" and "Build and push
  (alpine)" steps but I have no solution to propose for now;
- there are some linting changes (using prettier).

WIP because CI needs to be tested against upstream repo:
- create a DOCKER_USER secret so that forks can test the CI (push);
- I am using token and 2FA on DockerHub so the update description step
  could not be tested.

Regarding the last step (Update DockerHub description), we may consider using this https://github.com/marketplace/actions/docker-hub-description.

### CODE REVIEW

**Comment by fauust** [comment]
You can verify that the build/push was successful here https://github.com/fauust/sh/runs/1467387436?check_suite_focus=true and here https://hub.docker.com/repository/docker/fauust/shfmt.

**Review by mvdan — COMMENTED** [review]
Impressive work! Nice to see that we can drop a significant amount of tricky code.

Could you undo the indentation changes? They seem unnecessary, and will make the change significantly smaller.

Also, is this still WIP given that you tested it?

File: .github/workflows/test.yml
Comment (ArturKlauser) [inline .github/workflows/test.yml:146]: You might also want to replace this hand-rolled code with https://github.com/peter-evans/dockerhub-description

**Review by ArturKlauser — APPROVED** [review]
Very nice. The support for buildx has come a long way since I put in that code here. If you have any doubts on the details of why certain pieces of my original code are there, I've written up a [post](https://nexus.eddiesinentropy.net/2020/01/12/Building-Multi-architecture-Docker-Images-With-Buildx/) that might help you.

File: .github/workflows/test.yml
Comment (ArturKlauser) [inline .github/workflows/test.yml:111]: Ideally this would also be reorganized. What the current flow does is:
* build and push the image to dockerhub
* test the image from dockerhub to check if it works
* if there is a problem, an error is reported, however, the bogus image is already uploaded to dockerhub

It seems that there is built-in support now that makes it easy to [spin up a local temporary docker registry for testing purposes](https://github.com/docker/build-push-action#local-registry). So the better approach would be:

* build and push the image to a local temporary docker registry
* test the image from the local docker registry to check if it works
* if there is a problem report an error and quit
* if the tests were successful, push the image to dockerhub

File: .github/workflows/test.yml
Comment (fauust) [reply .github/workflows/test.yml:146]: Yes, fully agree, that's why I also suggested it in above ;-)

To be honest, I do not like so much the adherence to GitHub that this PR brings and you code was very instructive and useful to understand the workflow! I'll definitively take a look at your post.

Anyway, in the long run, this PR will help maintenance so I guess it's a good step forward :-)

File: .github/workflows/test.yml
Comment (fauust) [reply .github/workflows/test.yml:111]: Again, yes! Fully agree. Let's try to implement this in a next PR. Also, I would suggest splitting the workflows in multiple files to facilitate maintenance.

**Comment by fauust** [comment]
> Could you undo the indentation changes? They seem unnecessary, and will make the change significantly smaller.
Sure I will! But hey, that's sounds reasonable to me to lint the yml files in a shell linting project ;-)
Maybe in a future PR then...

> Also, is this still WIP given that you tested it?
The only thing that is missing is that you create a secret (that is not secret at all) but will facilitate GitHub action contributions:
Secret in your case is `DOCKER_USER = mvdan`

**Comment by ArturKlauser** [comment]
Suggestion regarding the required DOCKER_USER secret (and others): It's always a pain for contributors to reverse-engineer the implicit requirements of CI/CD flows. Ideally the CI file would allow you to specify which secrets are necessary and throw a clear error if one is missing. Could we approximate that with a step in the file that just checks if all required secrets are non-empty and throw an error if one is missing?

**Comment by fauust** [comment]
@ArturKlauser that's a good idea, how would you do that?

I thought that the GitHub action CI message would be pretty clear if it does not find:
```yaml
DOCKER_USER: ${{ secrets.DOCKER_USER }}
DOCKER_PASSWORD: ${{ secrets.DOCKER_PASSWORD }}
```
or would it just break later when it tries to connect to the DockerHub, I must admit I didn't check this eventuality.

**Comment by ArturKlauser** [comment]
> I thought that the GitHub action CI message would be pretty clear ...

Unfortunately that doesn't seem to be the case. The assignment of the ENV from non-existent secret fails silently and the actual failure happens when you're trying to use the ENV.

https://github.com/ArturKlauser/sh/runs/1468098862?check_suite_focus=true

```
...
Login to DockerHub
Run docker/login-action@v1
Error: Username and password required
...
```

**Comment by fauust** [comment]
@mvdan It should be OK now, let me know if I should target another test branch.
The only step that was not tested is the last one, "Update DockerHub description".

**Comment by mvdan** [comment]
Thank you for the effort! I'll take a look this week.

**Comment by fauust** [comment]
@mvdan as discussed in https://github.com/mvdan/sh/issues/631 with @ArturKlauser, I have prepared a new [branch](https://github.com/fauust/sh/tree/build-and-test-locally-first) that includes his suggestions.

If you prefer to go with a big single PR, we can forget this one and I will submit a new PR (that will introduce more changes in one shot). If you prefer multiple small PR, then I will open the new one as soon as this one is merged (but no urgency at all).

Let me know how you want to proceed...

**Comment by mvdan** [comment]
I definitely prefer smaller PRs :) reviewing now.

File: .github/workflows/test.yml
Comment (mvdan) [inline .github/workflows/test.yml:1]: can we remove this too? it seems like an artifact of the formatter you used at one point

File: .github/workflows/test.yml
Comment (mvdan) [inline .github/workflows/test.yml:58]: I found the older syntax easier to maintain to be honest; adding or deleting an item would just affect that line, and it was easy to review. I imagine this list will get longer, too.

File: .github/workflows/test.yml
Comment (mvdan) [inline .github/workflows/test.yml:102]: why do we need this id?

**Review by mvdan — APPROVED** [review]
Just minor nits, LGTM.

File: .github/workflows/test.yml
Comment (fauust) [reply .github/workflows/test.yml:1]: sure (this comes from [yaml specs](https://yaml.org/spec/1.2/spec.html#id2760395) but I have no strong opinion on whether it should be used or not. As far as I understand, this is not mandatory).

File: .github/workflows/test.yml
Comment (fauust) [reply .github/workflows/test.yml:58]: Agree, I also prefer the old syntax, but this seems how it should be used: https://github.com/docker/build-push-action#multi-platform-image

I can check if the following could work if you want:
```yaml
  DOCKER_PLATFORMS: |
    linux/386,
    linux/amd64,
    linux/arm/v7,
    linux/arm64/v8,
    linux/ppc64le
```

File: .github/workflows/test.yml
Comment (fauust) [reply .github/workflows/test.yml:102]: Not used, was part of a previous test, good catch!

File: .github/workflows/test.yml
Comment (ArturKlauser) [reply .github/workflows/test.yml:57]: The way that the previous version of the code did that is by programmatically converting the multi-line (space-separated) list into the right comma-separated format `--platform ${DOCKER_PLATFORMS// /,}`. I'm not sure if you can still do that directly where you need it, since `platforms: ${{ env.DOCKER_PLATFORMS }}` probably doesn't allow shell expansion. Alternatively, you could make it part of a `run` step and put it into a separate environment variable, like you do e.g. for setting up `$DOCKER_BASE`.

Comment (fauust) [reply .github/workflows/test.yml:57]: Sure, and we already have a run that defines some env variables (`Set up image tag`) so, I am happy with both solutions, let me know @mvdan what you prefer.

Comment (mvdan) [reply .github/workflows/test.yml:57]: I do prefer keeping the newline-separated list, if it is just one extra line of shell :)

Comment (fauust) [reply .github/workflows/test.yml:57]: ack

**Comment by fauust** [comment]
Voilà, @mvdan don't forget to create the DOCKER_USER secret and this should be ready to merge

**Comment by mvdan** [comment]
Thanks for all this work! Created the DOCKER_USER=mvdan secret, and merging.

**Comment by mvdan** [comment]
Also, if either of you would like to contribute to this project more actively, just let me know and I'd be happy to discuss giving write access to test CI/CD and be able to get stuff merged without requiring my approval.

**Comment by fauust** [comment]
Green, cool!

Yes, I am a big fan of this project, so if I still have extra time in the next month, I would like to contribute more.

The following are some ideas:
- finalize https://github.com/fauust/sh/tree/build-and-test-locally-first;
- implement pre-commit tests like this https://github.com/fauust/docker-systemd/runs/1467025202?check_suite_focus=true.

The first one was suggested by @ArturKlauser and is almost ready (we need to verify that we are re-using the cache efficiently). The second one is because I discovered that `cmd/shfmt/docker-entrypoint.sh` is not linted by shfmt and I think pre-commit.com project add a very elegant and easy to maintain way of not missing those things.

**Comment by mvdan** [comment]
> finalize https://github.com/fauust/sh/tree/build-and-test-locally-first;

sounds good!

> The second one is because I discovered that `cmd/shfmt/docker-entrypoint.sh` is not linted by shfmt and I think that the pre-commit.com project adds a very elegant and easy to maintain way of not missing those things.

I don't have anything against pre-commit per se, but I think that's a bit overkill. CI could just do something like `shfmt -w .; git diff --exit-code`, which will fail if there are any changes.

**Comment by mvdan** [comment]
In any case, the offer to give access to the repo is open; just reply here or ping me in the #shell channel at [the gophers slack](https://invite.slack.golangbridge.org/).

**Comment by ArturKlauser** [comment]
Thanks for the offer @mvdan, but I don't currently have time to pay enough attention here, so having you look over any changes before merging is a good thing.

**Comment by fauust** [comment]
> I don't have anything against pre-commit per se, but I think that's a bit overkill. CI could just do something like `shfmt -w .; git diff --exit-code`, which will fail if there are any changes.

Sure in this particular case yes, but it depends on how many linter/verifier you want to use in a repo. I see also markdown, yaml and maybe shellcheck pb. Anyway, maybe we should discuss this in an issue or a PR to have more input from other users.

**Comment by fauust** [comment]
Daniel Martí <notifications@github.com>,
10/12/2020 – 07:26:48 (-0800):

> In any case, the offer to give access to the repo is open; just reply
> here or ping me in the #shell channel at [the gophers
> slack](https://invite.slack.golangbridge.org/).

Great, so I am interested. Thanks!

**Comment by mvdan** [comment]
Sent an invite :)

**Pull request merged**
